astro/src/content/blog/forgerock-password-stroge.mdx

```diff
@@ -0,0 +1,143 @@
+
+Why quickly? Because you want to make sure your user can get through the authentication process quickly. They'll:
+
+* present their password
+* you'll add in the salt
+* you'll hash it one thousand times
+* you'll compare the computed value with the stored value
+
+Why not too quickly? Because if an attacker has a program submitting millions of passwords from a list, you want to slow them down and make it painful. For example, waiting 0.1 seconds when you are submitting one password is fine and won't be noticed. Waiting for one million passwords when each takes 0.1 seconds is more than a day.
+
+Record the number of iterations when you store the hash so that you can compare the correct computed password hash to the one stored.
+
+## How Do I Use It To Migrate Users?
+
+Secure identity and access management systems never store user passwords. They only store the hash. If they used a salted hash, they typically store the salt as well. By transferring the hash, the salt and the number of iterations to another system, you can derive the same hash if you are using the same hashing algorithm. 
+
+When a user enters their password on your site, apply the salt and apply the same hashing algorithm the same number of times that was used in the old system. Then compare the hash you get with the stored hash. They will match if the same password was used. 
+
+This is how you are able to migrate a user **without ever knowing the actual password**.
+
+## How Do I Decipher ForgeRock's Hash?
+Okay, now that you have the background and understand password hashing, let's dig into migrating ForgeRock password hashes.
+
+Password hash and salt storage will vary between implementations. One system may store the password and hash in different fields in a database and another may store them together in one field. Hashing algorithms can vary as well and could even vary between users. It is important to know these implementation details of the system you are getting the hashes from. By default, ForgeRock uses the PBKDF2-HMAC-SHA256 encryption with their Cloud Development Kit so that will be used in this article.
+
+<Aside type="note">
+If your implementation is using a different hashing function, ForgeRock has [documented their password storage schemes](https://backstage.forgerock.com/knowledge/kb/article/a44757687).
+</Aside>
+
+ForgeRock stores that hash in the following way: 
+
+`"{PBKDF2-HMAC-SHA256} ";" <iterations> ":" base64(<digest> <salt>)` where:
+
+- `{PBKDF2-HMAC-SHA256}` is a literal describing the hashing function used.
+- `";"` is a literal separator
+- `<iterations>` is the [iterations defined above](#iterations)
+- `":"` is a literal separator
+- `base64(<digest> <salt>)` holds the password hash and salt and is base64 encoded
```

Comment:
This `base64` is different casing than the `Base64` string you use below.

Reply:
I was in the process of changing probably while you were reviewing. Based on our scrum discussion, they should all be lower case now.